# Comparing Results and Runtime of Common Differential Gene Expression Analysis Tools

#### Sahej Randhawa

## Abstract

We sought to compare two different approaches to differential gene expression analysis: an alignment-based approach using the cufflinks library and a k-mer counting approach using kallisto and sleuth. We conducted this analysis in order to draw conclusions about when it would be best to use each of these tools and to summarize their strengths and limitations, especially regarding speed and accuracy. We compared these by performing each set of analysis comparing downsampled RNA-seq data of tail tip and tail base cells of the _Anolis carolinensis_ lizard. Kallisto+sleuth reported 1621 differentially expressed transcripts while cufflinks reported 42. While the downsampling prevented us from making a claim about accuracy, we did observe that the cufflinks approach was more time-intensive than kallisto but was able to output information about more functional units of the genome and was easier to perform correctly with limited data and replicates.

## Introduction

With the advent of RNA-seq data, differential gene expression analysis has become an essential approach to understanding the underlying expression patterns and genetic mechanisms that causes two experimental conditions to differ. These conditions can be different due to variables like time, stimuli like disease conditions, and the sample’s origins in terms of cell type and parent organism. To conduct this analysis, RNA-seq data is gathered by isolating the RNA in question, such as mRNA transcripts, reverse-transcribing it, and generating reads through a next-gen sequencing process. This reads data enables an estimation of transcript abundance and therefore of gene expression. Differential gene expression analysis seeks to compare these estimations across samples to identify the transcripts and genes that are expressed more in one sample, with “more” being defined by a threshold for statistical significance. An example of past research with findings based on differential gene expression, Hutchins et al. sought to analyze the different tail segments of the _Anolis carolinensis_ lizard to understand genetic patterns and mechanisms for limb regeneration [1]. This study used two pathways for conducting the differential gene expression analysis, thereby inspiring the biological question for this project, described next. 

For this type of analysis, the approaches and tools used can significantly impact the results, and we sought to compare the results of two different approaches, the first being the widely-used cufflinks library and the alignment-based algorithm it uses and the second being the k-mer counting based implemented by kallisto and its differential gene expression analysis companion sleuth. This query is important as these two approaches are significantly different and robustly assessing their strengths and limitations is critical to performing accurate and efficient differential expression analysis. This query is especially important because the alignment-based approach and the cufflinks library specifically are well-established, standard tools for this analysis but methods like kallisto are relatively more recent and claiming to be superior in terms of accuracy and speed; thus, the question at the center of this project aids in the conversation regarding which tool and approach is best overall or in specific cases.

As a background on these tools, their different approaches exist as it is difficult to measure and distinguish expression of genes merely from reads and fragments of transcriptomic information. Cufflinks solves this problem through relying on an alignment of the reads to a reference genome and then creating a weighted bipartite graph to represent fragments and finding paths in this graph to form transcripts [2]. This enables estimation of the abundance of the transcripts and thus of the expression of the genes they correspond to, specifically through using a model estimating abundance as a function of fragment count and splicing structure of the fragments based on the alignment information [3]. The authors of this tool contend that this model is superior to simpler common models like those based on a Poisson distribution [3]. This approach has been utilized widely, including in the aforementioned Hutchins paper [1], and is especially relied on for its ability to discover novel transcripts. 

By contrast, kallisto’s approach involves a pseudoalignment, which involves connecting reads to transcripts they could belong to but not on an exact base-level alignment as with cufflinks’ approach [4]. As this is accomplished through hashing on and counting k-mers, this approach enables faster but potentially inaccurate performance without countermeasures. As a result, kallisto takes the pseudoalignment information and creates a De Bruijn graph based on the k-mers where a path represents pseudoalignment of the k-mers into a transcript. The authors of this approach contend that these steps lead to not only faster analysis but also more accurate results, as its lack of a true alignment step gives this approach resistance to sequencing errors affecting bases and its bootstrapping step consistently resamples and recalculates differences in abundance for a more accurate estimation.

## Methods
We began with RNA-seq data from the previously mentioned Hutchins et al study [1], choosing our samples to compare as the tail tip (S5 segment) and tail base (S1 segment) of one lizard (SRA: SRX642055 and SRX642051, respectively). As this data would be too large for our account quota, I downsampled this data to 20% of its original number of reads. This was accomplished by using fastq-dump (version 2.9.0) with the –split-files option since the data in given as paired reads and by seqtk’s sample function (version 1.2) with a random seed of 100 [6]. For the first sample used for downsampling, a fraction of 0.2 was used and the exact number of reads this corresponded to was passed into seqtk for downsampling the second sample, in order to keep the number of reads between the finals the same and controlled for. We also downloaded chromosome 1 of the reference genome cited by this paper from NCBI as a fasta file (with GenBank id CM000937.1) [7] as another form of downsampling to stay within our computing resources. Other data required for this project involve gene annotation and transcriptome files as GFF3 and fasta files, respectively, from the same NCBI genome page for this model organism [7]. To get the annotations for just chromosome 1, we filtered by using the script getChr1Annotate.py, searching annotations by the NCBI RefSeq id for this chromosome found on the NCBI page. Lastly, we performed fastqc analysis (version 0.11.7) [8], which gave us the read length, and its passing labels given obviated the need for trimming.

The cufflinks arm of this project was conducted as follows. First, we built an index of the reference genome and aligned the S1 and S2 sample’s reads to the reference genome using the HiSat2 aligner (version 2.1.0), as this aligner is designed for this task. [9] Besides default options pointing to input and output files, a critical option used was the --dta-cufflink to add the attributes to the SAM file output required by cufflinks later in this analysis. We then filtered out the unaligned reads and sorted the remaining mapped reads using samtools (version 1.5), which included use of the -F option being given 4 to correspond to the appropriate filtering out of unmapped reads [10]. Then, we used cufflinks’s gffread with the -T option to convert our annotation file from GFF3 to GTF format, giving us all the necessary files in the right formats to use the cufflinks library. We first used cufflinks to generate a transcriptome from each sample’s RNA-seq reads, using the -N and -u flags to perform the upper quartile norm and multiple read correction procedures that reweights reads that are multiply aligned for improved accuracy and performs an upper quartile normalization, which is useful for cases like this where abundance of reads and transcripts will be less due to downsampling. Then we used cuffmerge to create a single transcriptome from both sample’s output transcriptomes from cufflinks, using the reference genome for improved mapping and 3 threads for faster runtime. Finally, we used cuffdiff with 3 threads, the gene annotation file, and default significance cutoff values to perform the abundance estimation and differential gene expression analysis of our sample reads onto this transcriptome. To obtain the number of differentially expressed, we used the unix grep tool to count the number of occurrences of the keyword “yes” in the column indicating significance for the outputted tables for transcript isoforms and genes. 

The kallisto arm of this project was conducted as follows. As our downloaded reference transcriptome included transcripts from all chromosomes rather than just chromosome 1, we needed to find all the transcripts originating from chromosome 1 in order to fairly compare the results of this arm with the cufflinks arms. Thus, we came up with the following approach. We used HiSat2 again to align the reference transcriptome to the reference chromosome 1 genome. Filtering out unmapped reads again with samtools, we then used the samtools fasta command with 3 threads to take the mapped reads in sam format and convert back into a fasta format that could be used by kallisto as a transcriptome index. With this step performed, we used kallisto index to create a transcriptome index and de Bruijn graph, which utilized the default k-mer size of 31. Then, we used kallisto quant to perform the pseudoalignment and quantify abundance of k-mers within our reads. For this step, we used 100 bootstrap experiments and 3 threads, as specified in the run_kallisto.sh script. Lastly, we turned to sleuth to perform the differential gene expression with default significant cutoff values, as seen in the script sleuthResults.R. An important part of this step was duplicating our sample to “create” another replicate, as sleuth requires multiple replicates since it uses a calculation of variance for its estimations. We at first attempted to download and use a real biological replicate from the given dataset, but since we needed to randomly downsample these reads as well to negotiate the data quota, the similarity between the replicates disappeared and 0 differentially expressed genes were found. Thus, we turned to the duplication approach as aforementioned. Further discussion and explanation of this is included in results and discussion sections. To obtain the number of differentially expressed transcripts/genes, we merely needed to count the number of lines in the output file from the sleuth script.

## Results
Our fastqc results confirmed the SRA page information that the reads were indeed paired and of length 39 base pairs. Our fastqc reports are attached in the accompanying fastqcOutput directory and reported passing labels for all categories except Overrepresented Sequences due to the poly-T and poly-N sequences from the RNA-seq process, which were removed in our subsequent steps such as in kallisto. The per base sequence quality graphs are shown below in Figure 1 for the S1 sample and Figure 2 for the S5 sample. It also confirmed that our downsampling reduced each sample to 5,059,449 paired reads.

#### Figure 1: S1 Sample's Fastqc Per Base Sequence Quality Graph
| a) Forward Reads | b) Reverse Reads |
|---------------------|------------------|
| ![alt text](https://github.com/cse185-sp18/cse185-final-project-sahejr/blob/master/fastqcOutput/S1_1.jpg "Figure 1a") | ![alt text](https://github.com/cse185-sp18/cse185-final-project-sahejr/blob/master/fastqcOutput/S1_2.jpg "Figure 1b") |

#### Figure 2: S5 Sample's Fastqc Per Base Sequence Quality Graph
| a) Forward Reads | b) Reverse Reads |
|---------------------|------------------|
| ![alt text](https://github.com/cse185-sp18/cse185-final-project-sahejr/blob/master/fastqcOutput/S5_1.jpg "Figure 2a") | ![alt text](https://github.com/cse185-sp18/cse185-final-project-sahejr/blob/master/fastqcOutput/S5_2.jpg "Figure 2b") |

This project’s alignment arm possessed 12.59% and 13.39% alignment rates for the S1 and S5 samples, respectively, to just the chromosome 1 reference genome, a result we deemed reasonable for further analysis. After performing our cuffdiff command for calculating differential gene expression, we obtained the output seen in the attached directory cuffDiffOut, which contains files with format and content as specified by cuffdiff [5]. The core results can be found in the isoform_exp.diff and gene_exp.diff files, which carry the id’s, expression values for each sample, q and p-values, and a yes/no declaration of statistical significance in difference of expression for each transcript and gene observed. From this arm of the analysis, we found that 42 transcripts were differentially expressed to a statistically significant level. For the kallisto arm of the analysis, the generation of the transcriptome for chromosome 1 seemed successful with 3930 transcripts used to form it. Ultimately, sleuth reported 1621 transcripts that were differentially expressed to a statistically significant level. Information analogous to the cuffdiff output for transcripts can be found in the attached file final_sleuth_results.tab. Thus, as observed from these values, there are significantly more differentially expressed transcripts reported by kallisto and sleuth than from the alignment approach with the cufflinks library. Analysis on how this result and the steps taken to produce these results comment on the performance of these two approaches can be found next in Discussion.

#### Table 2: Differential Transcript Expression Results for Both Analysis Arms
| Alignment (cufflinks) | k-mers and pseudoalignment (kallisto+sleuth)|
|--|--|
| 42  | 1621 |

## Discussion
At first glance, our results indicate that the kallisto and sleuth approach has a significantly lower threshold for determining a transcript as differentially expressed, potentially alluding at greater sensitivity and accuracy for the k-mer counting based kallisto method. However, because of our downsampling to just one chromosome and random reads from the given dataset, our data was not the same as that paper’s and we lacked a ground truth to compare the results against. Thus, with regard to our biological question, we were not able to definitively pinpoint which tool was more accurate, just which tool had that lower threshold for declaring something differentially expressed. Furthermore, our data quota and downsampling prevented us from using multiple replicates for either set of tools, which was especially detrimental for kallisto as it requires multiple replicates. As our attempt to add other replicates was also stymied by randomly downsampling creating too much variance between samples (since these similar samples now had a random set of transcripts that likely would not appear similar overall), our only solution was to duplicating our data. This resulted in the opposite problem, that of extremely low variance being inputted into kallisto, which explains its extremely low threshold for declaring something differentially expressed: the algorithm saw replicates within samples with identical data and thus any deviation between samples would be considered significant. 

However, these results do allow us to answer other questions about comparing these tools. With the given data restrictions, we were able to complete analysis that was much more standard and in line with our motivating paper when we used the alignment-based cufflinks approach, suggesting this approach is better given data restrictions. Furthermore, we saw that cuffdiff generated differential expression analysis for promoter, CDS sequences, and other units of the genome while kallisto only gave a perspective on transcripts, suggesting that cuffdiff gives more data for this type of analysis. Additionally, we were able conduct rough runtime measurements for each step of our work and qualitatively can state that the kallisto method was faster than the alignment-based cufflinks method, chiefly because of the lack of true alignment steps.

Our suggestion for further research would be to involve more replicates and more data for each sample in order to not suffer the aforementioned issues. This would also enable the results to be compared to those of a paper like Hutchins [1] for computing metrics like sensitivity and specificity. We also think that expanding the analysis to include performance metrics like time, storage space, and memory used for these processes would be critical in holistically evaluating these two tools. 

## Citations
1. Elizabeth Hutchins et al. (2014). Transcriptomic Analysis of Tail Regeneration in the Lizard Anolis carolinensis Reveals Activation of Conserved Vertebrate Developmental and Repair Mechanisms. PLoS ONE 9(8): e105004. https://doi.org/10.1371/journal.pone.0105004.

2. Cole Trapnell, Brian Williams, Geo Pertea, Ali Mortazavi, Gordon Kwan, Jeltje van Baren, Steven Salzberg, Barbara Wold, Lior Pachter (2010). Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation. Nature Biotechnology, 2010. doi:10.1038/nbt.1621

3. Cole Trapnell, David Hendrickson, Martin Sauvageau, Loyal Goff, John L. Rinn, Lior Pachter (2012). Differential analysis of gene regulation at transcript resolution with RNA-seq. Nature Biotechnology, 2012. doi:10.1038/nbt.2450.

4. Nicolas L Bray, Harold Pimentel, Páll Melsted and Lior Pachter, Near-optimal probabilistic RNA-seq quantification, Nature Biotechnology 34, 525–527 (2016), doi:10.1038/nbt.3519

5. Harold Pimentel, Nicolas L Bray, Suzette Puente, Páll Melsted and Lior Pachter, Differential analysis of RNA-seq incorporating quantification uncertainty, in press.

6. Heng Li (200). Seqtk. https://github.com/lh3/seqtk.

7. Genome [Internet]. Bethesda (MD): National Library of Medicine (US), National Center for Biotechnology Information; [1988]. Taxonomy ID 28377, Anolis carolinensis (green anole); [cited 2019 June 10]. Available from: https://www.ncbi.nlm.nih.gov/genome/?term=txid28377[orgn]

8. Andrews S. (2010). FastQC: a quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc.

9. Kim D, Langmead B and Salzberg SL. HISAT: a fast spliced aligner with low memory requirements. Nature Methods, 2015. https://ccb.jhu.edu/software/hisat2/manual.shtml.

10. Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009). The Sequence alignment/map (SAM) format and SAMtools. Bioinformatics, 25, 2078-9. [PMID: 19505943].
